### PR TITLE
Fw vpn sharing

### PR DESCRIPTION
**Pull request by ashutosh-mishra**

### CODE REVIEW

**Comment by ashutosh-mishra** [comment]
@njagadish @rajendramachani

File: gbpservice/neutron/services/servicechain/plugins/ncp/node_drivers/nfp_node_driver.py [hunk 1]
```diff
@@ -283,6 +291,40 @@ def get_plumbing_info(self, context):
...
 
+    def _check_for_fw_vpn_sharing(self, context, service_type):
+        shared_svc_type = pconst.FIREWALL
+        if service_type == pconst.FIREWALL:
+            shared_svc_type = pconst.VPN
+        return self._is_service_type_in_chain(context, shared_svc_type)
+
+    def _get_service_type(self, profile):
+        if profile['service_type'].endswith('_HA'):
+            service_type = profile['service_type'][:-3]
```
Comment (rajendramachani) [inline gbpservice/neutron/services/servicechain/plugins/ncp/node_drivers/nfp_node_driver.py]: HA is not yet there right. We don't need this if block. We can have this function for place holder and return profile['service_type']

File: gbpservice/nfp/Automation-Scripts/NFP_SCRIPTS/templates/vpn.template [hunk 2]
```diff
@@ -79,9 +79,9 @@
                 },
                 "mtu": 1500,
                 "name": "site_to_site_connection1",
-                "peer_address": "192.168.102.89",
-                "peer_cidrs": ["12.0.0.0/24"],
-                "peer_id": "12.0.0.3",
+                "peer_address": "192.168.102.117",
+                "peer_cidrs": ["19.0.0.0/24"],
+                "peer_id": "192.168.102.117",
```
Comment (sdharav) [inline gbpservice/nfp/Automation-Scripts/NFP_SCRIPTS/templates/vpn.template]: keeping both peer_address and peer_id same does not work in case of heterogeneous vendors, it is good to keep peer_id as the fixed ip of the peer.

File: gbpservice/nfp/service_vendor_agents/vyos/oc_config_server/static_ip.py [hunk 3]
```diff
@@ -2,6 +2,7 @@
 import logging
 import netifaces
 import time
+from netifaces import AF_LINK
```
Comment (rajendramachani) [inline gbpservice/nfp/service_vendor_agents/vyos/oc_config_server/static_ip.py]: Duplicate import statement

File: gbpservice/neutron/services/servicechain/plugins/ncp/node_drivers/nfp_node_driver.py [hunk 4]
```diff
@@ -34,6 +34,7 @@
 from gbpservice.neutron.services.servicechain.plugins.ncp import (
     exceptions as exc)
 from gbpservice.neutron.services.servicechain.plugins.ncp import driver_base
+from gbpservice.neutron.services.servicechain.plugins.ncp import model
```
Comment (rajendramachani) [inline gbpservice/neutron/services/servicechain/plugins/ncp/node_drivers/nfp_node_driver.py]: We can import this as ncp_model which will be readable in the code.

File: gbpservice/neutron/services/servicechain/plugins/ncp/node_drivers/nfp_node_driver.py [hunk 5]
```diff
@@ -273,6 +274,13 @@ def get_plumbing_info(self, context):
 
         if service_type in [pconst.FIREWALL, pconst.VPN]:
             plumbing_request['plumbing_type'] = 'gateway'
+            # plumber will return stitching network PT instead of consumer
+            # as chain is instantiated while creating provider group.
+            if (self._check_for_fw_vpn_sharing(context, service_type) and
+                    service_type == pconst.VPN):
+                LOG.info(_("Not requesting plumber for PTs for service type "
+                           "%s") % service_type)
+                return False
```
Comment (rajendramachani) [inline gbpservice/neutron/services/servicechain/plugins/ncp/node_drivers/nfp_node_driver.py]: Why return value is False
I think we need to return empty dict {} here, Isn't it ?

File: gbpservice/neutron/services/servicechain/plugins/ncp/node_drivers/nfp_node_driver.py [hunk 6]
```diff
@@ -576,11 +638,21 @@ def _get_service_targets(self, context):
         # Bug with NCP. For create, its not setting service targets in context
         if not service_targets:
             service_targets = context.get_service_targets(update=True)
-        for service_target in service_targets:
-            if service_target.relationship == 'consumer':
-                consumer_service_targets.append(service_target)
-            elif service_target.relationship == 'provider':
-                provider_service_targets.append(service_target)
+        if (not service_targets and service_type in
+            [pconst.FIREWALL, pconst.VPN]):
```
Comment (rajendramachani) [inline gbpservice/neutron/services/servicechain/plugins/ncp/node_drivers/nfp_node_driver.py:642]: According to get_plumbing_info, service targets are created only for FW right. So, here service type is VPN alone. I think no need to check for FW.

File: gbpservice/nfp/orchestrator/drivers/orchestration_driver.py [hunk 7]
```diff
@@ -410,11 +435,18 @@ def select_network_function_device(self, devices, device_data):
...
+                                                   network_handler))
+        service_type = device_data['service_details']['service_type']
         for device in devices:
             if (
                 (device['interfaces_in_use'] + hotplug_ports_count) <=
                 self.maximum_interfaces
             ):
+                if (service_type.lower() == nfp_constants.VPN.lower() and
+                    service_type in device_service_types_map[device['id']]):
+                    continue
```
Comment (rajendramachani) [inline gbpservice/nfp/orchestrator/drivers/orchestration_driver.py:453]: Can we add a comment for this condition check

File: gbpservice/nfp/orchestrator/modules/device_orchestrator.py [hunk 8]
```diff
@@ -432,13 +432,30 @@ def _delete_network_function_device_db(self, device_id, device):
...
+        nfi_filters = {'network_function_device_id': [device_id]}
+        network_function_instances = (
+            self.nsf_db.get_network_function_instances(self.db_session,
+                                                       nfi_filters))
+        network_function_ids = [nf['network_function_id']
+                for nf in network_function_instances]
+        network_functions = (
+            self.nsf_db.get_network_functions(self.db_session,
+                {'id': network_function_ids}))
+        return network_functions, network_function_instances
```
Comment (rajendramachani) [inline gbpservice/nfp/orchestrator/modules/device_orchestrator.py]: The part of the return value network_function_instances is not used
We can just return network_functions

File: gbpservice/nfp/orchestrator/modules/device_orchestrator.py [hunk 9]
```diff
@@ -432,13 +432,30 @@ def _delete_network_function_device_db(self, device_id, device):
...
                                                 self.db_session, filters)
         for device in network_function_devices:
             mgmt_port_id = device.pop('mgmt_port_id')
             mgmt_port_id = self._get_port(mgmt_port_id)
             device['mgmt_port_id'] = mgmt_port_id
+
+            network_functions, network_function_instances = (
+                    self._get_network_function_info(device['id']))
+            device['network_functions'] = network_functions
+            # device['network_function_instances'] = network_function_instances
```
Comment (rajendramachani) [inline gbpservice/nfp/orchestrator/modules/device_orchestrator.py]: Remove unwanted code

File: gbpservice/nfp/orchestrator/modules/service_orchestrator.py [hunk 10]
```diff
@@ -750,10 +750,32 @@ def delete_user_config(self, event):
...
 
+    def _get_network_function_instance_for_multi_service_sharing(self,
+                                                                 port_info):
+        network_function_instances = (
+            self.db_handler.get_network_function_instances(self.db_session,
+                                                       filters={}))
+        provider_port_id = None
+        for port in port_info:
+            if port['port_classification'] == 'provider':
+                provider_port_id = port['id']
```
Comment (rajendramachani) [inline gbpservice/nfp/orchestrator/modules/service_orchestrator.py:761]: break statement is needed right after this assignment

File: gbpservice/nfp/service_plugins/firewall/nfp_fwaas_plugin.py [hunk 11]
```diff
@@ -20,3 +25,30 @@ def __init__(self):
                 cfg.CONF.api_extensions_path + ':' + ext_path)
 
         super(NFPFirewallPlugin, self).__init__()
+
+    def _get_routers_for_create_firewall(self, tenant_id, context, firewall):
```
Comment (rajendramachani) [inline gbpservice/nfp/service_plugins/firewall/nfp_fwaas_plugin.py:30]: Why this method is needed, and who calls this.
We can add few comments here for maintenance.

File: gbpservice/neutron/services/servicechain/plugins/ncp/node_drivers/nfp_node_driver.py [hunk 12]
```diff
@@ -273,6 +274,13 @@ def get_plumbing_info(self, context):
 
         if service_type in [pconst.FIREWALL, pconst.VPN]:
             plumbing_request['plumbing_type'] = 'gateway'
+            # plumber will return stitching network PT instead of consumer
+            # as chain is instantiated while creating provider group.
+            if (self._check_for_fw_vpn_sharing(context, service_type) and
+                    service_type == pconst.VPN):
```
Comment (rajendramachani) [inline gbpservice/neutron/services/servicechain/plugins/ncp/node_drivers/nfp_node_driver.py:280]: We can add comment here to explain about this check. about FW and VPN sharing.

File: gbpservice/nfp/configurator/drivers/firewall/vyos/vyos_fw_driver.py [hunk 13]
```diff
@@ -364,7 +364,8 @@ def configure_routes(self, context, resource_data):
                "primary service at: %r" % mgmt_ip)
         LOG.info(msg)
         try:
-            resp = requests.post(url, data=data, timeout=self.timeout)
+            routes_timeout = 60
```
Comment (njagadish) [inline gbpservice/nfp/configurator/drivers/firewall/vyos/vyos_fw_driver.py:367]: Why is the routes timeout changed?

Comment (ashutosh-mishra) [inline gbpservice/nfp/configurator/drivers/firewall/vyos/vyos_fw_driver.py:367]: There is a big issue w.r.t timeouts.
Observed that by default timeout is 30 secs, This particular rest call without sharing took 27~29 secs, with sharing taking 34~36 secs, hence failing.
For now increased timeout to large value 60 secs, but we need to validate all timeouts properly.

Comment (ashutosh-mishra) [inline gbpservice/nfp/configurator/drivers/firewall/vyos/vyos_fw_driver.py:367]: Same applies for asav drivers too.

File: gbpservice/nfp/orchestrator/drivers/orchestration_driver.py [hunk 14]
```diff
@@ -36,7 +37,10 @@
 
 def _set_network_handler(f):
     def wrapped(self, *args, **kwargs):
-        device_data = args[0]
+        if type(args[0]) == dict:
```
Comment (njagadish) [inline gbpservice/nfp/orchestrator/drivers/orchestration_driver.py:40]: Depending on data type to pick argument selectively is a risk. Can't this be done differently?

File: gbpservice/nfp/orchestrator/drivers/orchestration_driver.py [hunk 15]
```diff
@@ -52,7 +56,7 @@ class OrchestrationDriver(object):
     is launched for each Network Service Instance
     """
     def __init__(self, config, supports_device_sharing=True,
-                 supports_hotplug=True, max_interfaces=5):
+                 supports_hotplug=True, max_interfaces=10):
```
Comment (njagadish) [inline gbpservice/nfp/orchestrator/drivers/orchestration_driver.py:59]: Isn't max_interfaces count change based on vendor?

File: gbpservice/neutron/services/servicechain/plugins/ncp/node_drivers/nfp_node_driver.py [hunk 16]
```diff
@@ -283,6 +294,37 @@ def get_plumbing_info(self, context):
...
+        else:
+            current_specs = context.relevant_specs
+            service_profiles = []
+            for spec in current_specs:
+                filters = {'id': spec['nodes']}
+                nodes = context.sc_plugin.get_servicechain_nodes(
+                    context.plugin_context, filters)
+                for node in nodes:
+                    service_profiles.append(node['service_profile_id'])
+            service_type_ha = "%s%s" %(service_type, "_HA")
```
Comment (njagadish) [inline gbpservice/neutron/services/servicechain/plugins/ncp/node_drivers/nfp_node_driver.py:320]: Why HA is coming into picture for the sharing feature?

File: gbpservice/neutron/services/servicechain/plugins/ncp/node_drivers/nfp_node_driver.py [hunk 17]
```diff
@@ -273,6 +274,16 @@ def get_plumbing_info(self, context):
 
         if service_type in [pconst.FIREWALL, pconst.VPN]:
             plumbing_request['plumbing_type'] = 'gateway'
+            # plumber will return stitching network PT instead of consumer
+            # as chain is instantiated while creating provider group.
+            if (self._check_for_fw_vpn_sharing(context, service_type) and
```
Comment (njagadish) [inline gbpservice/neutron/services/servicechain/plugins/ncp/node_drivers/nfp_node_driver.py:279]: Will the node create order always FW and followed by VPN? Is this the reason why we always skip interface create for VPN?

File: gbpservice/neutron/services/servicechain/plugins/ncp/node_drivers/nfp_node_driver.py [hunk 18]
```diff
@@ -273,6 +274,16 @@ def get_plumbing_info(self, context):
...
+                    service_type == pconst.VPN):
+                # For fw, vpn sharing, request pts only once.
+                # Request pts only for firewall, if both fw, vpn is in chain,
+                # and current service type is vpn, dont request pts.
+                LOG.info(_("Not requesting plumber for PTs for service type "
+                           "%s") % service_type)
```
Comment (njagadish) [inline gbpservice/neutron/services/servicechain/plugins/ncp/node_drivers/nfp_node_driver.py:285]: In case of FW+FW sharing, interfaces are not shared but only service VM? Is this correct?

File: gbpservice/nfp/Automation-Scripts/NFP_SCRIPTS/templates/vpn.template [hunk 19]
```diff
@@ -79,9 +79,9 @@
                 },
                 "mtu": 1500,
                 "name": "site_to_site_connection1",
-                "peer_address": "192.168.102.89",
-                "peer_cidrs": ["12.0.0.0/24"],
-                "peer_id": "12.0.0.3",
+                "peer_address": "192.168.102.117",
+                "peer_cidrs": ["19.0.0.0/24"],
```
Comment (njagadish) [inline gbpservice/nfp/Automation-Scripts/NFP_SCRIPTS/templates/vpn.template:83]: If this change is not needed, revert it.

File: gbpservice/nfp/configurator/drivers/firewall/vyos/vyos_fw_driver.py (hunk 13 above)
Comment (njagadish) [inline gbpservice/nfp/configurator/drivers/firewall/vyos/vyos_fw_driver.py:367]: As discussed, revert this change

**Comment by ashutosh-mishra** [comment]
Closed as raised new PR to https://github.com/oneconvergence/group-based-policy/pull/390

**Comment by ashutosh-mishra** [comment]
will address remaining review comments in next PR
